File: lib/domain/models/library.dart
class Library {
  const Library({
    required this.systemId,
    required this.systemName,
    required this.libKey,
    required this.libId,
    required this.shortName,
    required this.formalName,
    required this.address,
    required this.pref,
    required this.city,
    required this.category,
    this.url,
    this.tel,
    this.geocode,
  });

  final String systemId;
  final String systemName;
  final String libKey;
  final String libId;
  final String shortName;
  final String formalName;
  final String address;
  final String pref;
  final String city;
  final String category;
  final String? url;
  final String? tel;
  final String? geocode;

  factory Library.fromJson(Map<String, dynamic> json) {
    return Library(
      systemId: json['systemId'] as String,
      systemName: json['systemName'] as String,
      libKey: json['libKey'] as String,
      libId: json['libId'] as String,
      shortName: json['shortName'] as String,
      formalName: json['formalName'] as String,
      address: json['address'] as String,
      pref: json['pref'] as String,
      city: json['city'] as String,
      category: json['category'] as String,
      url: json['url'] as String?,
      tel: json['tel'] as String?,
      geocode: json['geocode'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'systemId': systemId,
      'systemName': systemName,
      'libKey': libKey,
      'libId': libId,
      'shortName': shortName,
      'formalName': formalName,
      'address': address,
      'pref': pref,
      'city': city,
      'category': category,
      'url': url,
      'tel': tel,
      'geocode': geocode,
    };
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Library &&
          runtimeType == other.runtimeType &&
          systemId == other.systemId &&
          libKey == other.libKey &&
          libId == other.libId;

  @override
  int get hashCode => Object.hash(systemId, libKey, libId);
}

File: src/domain/models/library.ts
export interface Library {
  readonly systemId: string;
  readonly systemName: string;
  readonly libKey: string;
  readonly libId: string;
  readonly shortName: string;
  readonly formalName: string;
  readonly address: string;
  readonly pref: string;
  readonly city: string;
  readonly category: string;
  readonly url?: string;
  readonly tel?: string;
  readonly geocode?: string;
}

function requireString(json: Record<string, unknown>, key: string): string {
  const value = json[key];
  if (typeof value !== 'string') {
    throw new Error(`Library.fromJson: missing or invalid required string "${key}"`);
  }
  return value;
}

function optionalString(json: Record<string, unknown>, key: string): string | undefined {
  const value = json[key];
  if (value === null || value === undefined) return undefined;
  if (typeof value !== 'string') {
    throw new Error(`Library.fromJson: invalid optional string "${key}"`);
  }
  return value;
}

export function libraryFromJson(json: Record<string, unknown>): Library {
  return {
    systemId: requireString(json, 'systemId'),
    systemName: requireString(json, 'systemName'),
    libKey: requireString(json, 'libKey'),
    libId: requireString(json, 'libId'),
    shortName: requireString(json, 'shortName'),
    formalName: requireString(json, 'formalName'),
    address: requireString(json, 'address'),
    pref: requireString(json, 'pref'),
    city: requireString(json, 'city'),
    category: requireString(json, 'category'),
    url: optionalString(json, 'url'),
    tel: optionalString(json, 'tel'),
    geocode: optionalString(json, 'geocode'),
  };
}

export function libraryToJson(lib: Library): Record<string, unknown> {
  return {
    systemId: lib.systemId,
    systemName: lib.systemName,
    libKey: lib.libKey,
    libId: lib.libId,
    shortName: lib.shortName,
    formalName: lib.formalName,
    address: lib.address,
    pref: lib.pref,
    city: lib.city,
    category: lib.category,
    url: lib.url ?? null,
    tel: lib.tel ?? null,
    geocode: lib.geocode ?? null,
  };
}

export function librariesEqual(a: Library, b: Library): boolean {
  return a.systemId === b.systemId && a.libKey === b.libKey && a.libId === b.libId;
}

export function libraryKey(lib: Library): string {
  return `${lib.systemId}|${lib.libKey}|${lib.libId}`;
}
